### PR TITLE
payment_authorize to new payment api

### PR DESCRIPTION
**Pull request by kbapt**

### CODE REVIEW

File: addons/payment_authorize/models/payment_token.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+from .authorize_request import AuthorizeAPI
+
+
+class PaymentToken(models.Model):
+    _inherit = 'payment.token'
+
+    authorize_profile = fields.Char(string='Authorize.net Profile ID', help='This contains the unique reference '
+                                    'for this partner/payment token combination in the Authorize.net backend')
+    provider = fields.Selection(string='Provider', related='acquirer_id.provider', readonly=False)
```
Comment (Feyensv) [inline addons/payment_authorize/models/payment_token.py]: Is this strange related field still needed ?
I find it quite strange to be able to modify the provider field of an acquirer from a token :D

File: addons/payment_authorize/static/src/js/payment_form.js [hunk 2]
```diff
@@ -1,160 +1,101 @@
...
+        _authorizeResponseHandler: function (acquirerId, response) {
             if (response.messages.resultCode === "Error") {
-                var errorMessage = "";
-                _.each(response.messages.message, function (message) {
-                    errorMessage += message.code + ": " + message.text;
-                })
-                acquirerForm.removeClass('d-none');
-                return self.displayError(_t('Server Error'), errorMessage);
+                let error = "";
+                response.messages.message.forEach((msg) => error += msg.code + ": " + msg.text + "\n");
```
Comment (Feyensv) [inline addons/payment_authorize/static/src/js/payment_form.js]: messages.message.forEach? strange but I suppose it works :D.

File: addons/payment_authorize/views/payment_authorize_templates.xml [hunk 3]
```diff
@@ -1,71 +1,34 @@
...
-         <template id="assets_frontend" inherit_id="web.assets_frontend">
-            <xpath expr="script[last()]" position="after">
-                <script type="text/javascript" src="/payment_authorize/static/src/js/payment_form.js"></script>
-            </xpath>
-        </template>
-    </data>
+        <xpath expr="link[last()]" position="after">
+            <link rel="stylesheet" type="text/scss" href="/payment_authorize/static/src/scss/payment_authorize.scss" />
+        </xpath>
+    </template>
```
Comment (Feyensv) [inline addons/payment_authorize/views/payment_authorize_templates.xml]: Specify the assets in a specific file ;).

File: addons/payment_authorize/views/payment_views.xml [hunk 4]
```diff
@@ -1,36 +1,45 @@
...
+                </group>
+            </xpath>
+            <xpath expr="//field[@name='check_validity']" position="attributes">
+                <attribute name="attrs">{'invisible': ['|', ('support_tokenization', '=', False), ('provider', '=', 'authorize')]}</attribute>
+            </xpath>
+            <xpath expr="//field[@name='country_ids']" position="after">
+                <label for="authorize_currency_ids" attrs="{'invisible': [('provider', '!=', 'authorize')]}" />
+                <div attrs="{'invisible': [('provider', '!=', 'authorize')]}">
+                   <field name="authorize_currency_ids" widget="many2many_tags" />
+                    <button class="oe_link" icon="fa-refresh" type="object" name="action_get_currencies" string="Get Supported Currencies" />
```
Comment (Feyensv) [inline addons/payment_authorize/views/payment_views.xml]: inconsistent indentation ^^

File: addons/payment_authorize/controllers/main.py [hunk 5]
```diff
@@ -1,89 +1,29 @@
...
-        res = {
-            'result': True,
-            'id': token.id,
-            'short_name': token.short_name,
-            '3d_secure': False,
-            'verified': True, #Authorize.net does a transaction type of Authorization Only
-                              #As Authorize.net already verify this card, we do not verify this card again.
+    @http.route(['/payment/authorize/get_acquirer_info'], type='json', auth='public')
+    def authorize_get_acquirer_info(self, acquirer_id):
+        acquirer_sudo = request.env['payment.acquirer'].sudo().browse(acquirer_id)
```
Comment (Feyensv) [inline addons/payment_authorize/controllers/main.py]: `.exists()` ?

**Review by Feyensv — COMMENTED** [review]
Quick look, can be for ANV during his review

File: addons/payment_authorize/models/payment_token.py (hunk 1 above)
Comment (kbapt) [reply addons/payment_authorize/models/payment_token.py]: It should be readonly :grimacing: , but it's used to hide the `authorize_profile` on the `payment.token` form view

File: addons/payment_authorize/controllers/main.py [hunk 6]
```diff
@@ -11,21 +11,36 @@ class AuthorizeController(http.Controller):
...
-    @http.route('/payment/authorize/payments', type='json', auth='public')
-    def authorize_payments(self, opaque_data, reference, access_token, partner_id):
+    @http.route('/payment/authorize/payment', type='json', auth='public')
+    def authorize_payment(self, reference, partner_id, access_token, opaque_data):
+        """ Make a payment request and handle the response.
+
+        :param str reference: The reference of the transaction
+        :param int partner_id: The partner making the transaction, as a `res.partner` id
+        :param str access_token: The access token used to verify the provided values
+        :param dict opaque_data: The payment nonce provided by Authorize.Net
```
Comment (Feyensv) [inline addons/payment_authorize/controllers/main.py]: nonce ?

Comment (kbapt) [reply addons/payment_authorize/controllers/main.py]: That's the name Authorize use for the single-use payment token in their doc :grimacing: 

https://developer.authorize.net/api/reference/features/acceptjs.html

Comment (Feyensv) [reply addons/payment_authorize/controllers/main.py]: Thanks, Antoine just explained it to me (and we replaced it by a more generic naming in the code ^^).

File: addons/payment_authorize/models/payment_transaction.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+            _logger.info("Authorize: " + _('error received whilst processing payment:\n%s', error))
+            self._set_error(error)
+
+    def _authorize_make_payment(self, opaque_data):
+        self.ensure_one()
+        authorize = AuthorizeAPI(self.acquirer_id)
+        if self.acquirer_id.capture_manually or self.operation == 'validation':
+            res = authorize.authorize(self.amount, self.reference, opaque_data=opaque_data)
+        else:
+            res = authorize.auth_and_capture(self.amount, self.reference, opaque_data=opaque_data)
```
Comment (Feyensv) [inline addons/payment_authorize/models/payment_transaction.py]: Why isn't the amount rounded here ? (it is rounded below when going through the "same" authorize route)

Comment (kbapt) [reply addons/payment_authorize/models/payment_transaction.py]: Seems to be an oversight, it should be rounded indeed

**Review by AntoineVDV — APPROVED** [review]
